[기본 레이아웃 컴포넌트] 코난(윤정민) 미션 제출합니다.

jariita opened the pull request:
# 🎯 1단계. 기본 레이아웃 컴포넌트

## 공통
- [x] 의미있는 커밋 메시지 작성
- [x] PR에 관련 없는 변경 사항이 포함 유무 확인

## Container 컴포넌트
- [x] `minWidth`, `maxWidth` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## Grid 컴포넌트
- [x] `rows`, `columns`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## Flex 컴포넌트
- [x] `direction`, `justify`, `align`, `gap` 속성 적용
- [x] npm 배포
- [x] README.md 작성
- [ ] 테스트 코드를 작성하였나요? (선택 사항)

## 📌 구현한 내용 설명

### 공통

구현한 모든 컴포넌트는 레이아웃만 결정한다 생각하여 width, height 속성은 주지 않았습니다.
하지만 속성이 부족하다 생각이 든다면 추가적인 속성을 적용할 수 있도록 props를 열어두었습니다.
as라는 속성을 통해 기본 태그인 `div`를 다른 태그로 변경할 수 있도록 적용하였습니다.

### Container 컴포넌트

> 화면의 폭에 따라 내용의 최대 폭을 제한하고 중앙에 배치하는 레이아웃을 만들기 위한 컴포넌트

내용의 최대폭과 최소폭만 결정할 수 있도록 하였고 중앙정렬을 하기위해 `margin: 0 auto` 속성을 사용하였습니다. 

### Grid 컴포넌트

> CSS Grid를 사용하여 자식 컴포넌트들을 격자 형태로 배열하는 컴포넌트

rows 속성을 통해 grid의 row 갯수를 강제할 수 있습니다.
columns 속성을 통해 grid의 column 갯수를 강제할 수 있습니다.
rowGap과 columnGap, gap 속성을 통해 grid의 간격을 조절할 수 있습니다.

### Flex 컴포넌트

> CSS Flexbox를 사용하여 자식 컴포넌트들을 유연하게 배열하는 컴포넌트

direction, wrap 속성을 통해 flex-flow를 통제할 수 있습니다.
justify, align 속성을 통해 flexbox를 정렬할 수 있습니다.
gap 속성을 통해 flexbox의 간격을 조절할 수 있습니다.

## 🔗 참고 링크

[npm 배포 페이지](https://www.npmjs.com/package/conan-layout-component)
[스토리북 배포 페이지](https://6507fe70aa120a5453701630-uaqgayiynj.chromatic.com/)

jariita committed:
chore: 패키지 설치

- react, typescript, webpack, babel, emotion, storybook

jariita committed:
chore: 프로젝트 환경 설정

- webpack, babel, tsconfig, gitignore

jariita committed:
chore: storybook 환경설정

jariita committed:
feat: Container 컴포넌트 구현

jariita committed:
feat: Container 컴포넌트 스토리북 작성

jariita committed:
feat: Grid 컴포넌트 구현

jariita committed:
feat: Grid 컴포넌트 스토리북 작성

jariita committed:
feat: Flex 컴포넌트 구현

jariita committed:
feat: Flex 컴포넌트 스토리북 작성

jariita committed:
docs: README.md 업데이트

jariita committed:
chore: chromatic 설치

jariita committed:
feat: index.ts 추가

woo-jk commented on README.md:
```diff
@@ -1,6 +1,124 @@
+| as?       | ElementType | Container 컴포넌트의 태그를 변경합니다. |
+| minWidth? | number      | Container의 최소 넓이 값입니다.         |
+| maxWidth? | number      | Container의 최대 넓이 값입니다.         |
+| children  | ReactNode   | Container의 자식 컴포넌트입니다.        |
```
README에 테이블 형식으로 프롭스 구성도를 보여주니 훨씬 이해하기 좋은 것 같아요~ md 파일을 잘 쓰시는군요👍

woo-jk commented on README.md:
```diff
@@ -1,6 +1,124 @@
+| wrap?      | string    | Flex 아이템들의 묶음(줄 바꿈)을 지정합니다. |
+| justify?   | string    | Flex direction의 수평축 정렬을 지정합니다.  |
+| align?     | string    | Flex direction의 수직축 정렬을 지정합니다.  |
+| gap        | number    | Flex 아이템들의 간격을 지정합니다.          |
```
실제 컴포넌트는 gap이 옵셔널한데 여기는 옵셔널 처리가 안되어있네요?

jariita replied:
✍️ **[옵셔널 + 기본값]** - [ca5282f](https://github.com/woowacourse/layout-component/pull/43/commits/ca5282f89478b135eec1b2f40d423ce744939805)와 [0cad914](https://github.com/woowacourse/layout-component/pull/43/commits/0cad914be63a3cd2dcc7d2632eddd944be76bb04)에서 수정하였습니다!

woo-jk commented on babel.config.js:
```diff
@@ -0,0 +1,7 @@
+module.exports = {
+  presets: [
+    ['@babel/preset-env'],
+    ['@babel/preset-react', { runtime: 'automatic', importSource: '@emotion/react' }],
```
각 컴포넌트의 css를 어디서 불러오나 했더니 바벨에서 이모션 설정을 해준거군요 ㄷㄷ

woo-jk commented on src/components/Flex.tsx:
```diff
@@ -0,0 +1,54 @@
+import { ComponentPropsWithoutRef, ElementType } from 'react';
+
+interface FlexProps extends ComponentPropsWithoutRef<ElementType> {
+  /**
+   * `Flex` 컴포넌트의 태그를 변경합니다.
+   * @default div
+   */
+  as?: ElementType;
+  /**
+   * `flex-direction` 스타일 프로퍼티를 정의합니다.
+   * @default row
+   */
+  direction?: string;
+  /**
+   * `flex-wrap` 스타일 프로퍼티를 정의합니다.
+   * @default wrap
+   */
+  wrap?: string;
+  /**
+   * `justify-content` 스타일 프로퍼티를 정의합니다.
+   * @default flex-start
+   */
+  justify?: string;
+  /**
+   * `align-items` 스타일 프로퍼티를 정의합니다.
+   * @default stretch
+   */
+  align?: string;
+  /**
+   * `gap` 스타일 프로퍼티를 정의합니다.
+   * @default 0
+   */
+  gap?: number;
```
프롭스별 주석을 상세히 달아주신 점 좋네요👍

한가지 팁을 드리자면, (이미 아실 수도 있습니다 ㅋㅋ) `JSDoc`을 학습 해보시고 적용해보는건 어떨까요?

<img width="819" alt="스크린샷 2023-09-21 오전 1 24 06" src="https://github.com/woowacourse/layout-component/assets/73513965/c073428a-a81c-404b-af73-7fa23e708430">

컴포넌트의 선언 위에 사진처럼 props 같은 요소들을 작성해주시고, 이 컴포넌트를 사용하는 곳에서 컴포넌트 위에 마우스를 올려보면? 

<img width="666" alt="스크린샷 2023-09-21 오전 1 24 14" src="https://github.com/woowacourse/layout-component/assets/73513965/464f1e76-22f4-4320-980d-9ad9b207e004">

작성한 주석을 바로 볼 수 있어요! jsdoc을 적극 활용하면 DX가 아주 상승하겠죠?? ㅎㅎ

jariita replied:
JSDoc에는 `@props`가 없던데 예시로 알려주신 것은 어떤 의도로 사용하는 방법인가요?!? 😭
(JSDoc에 없는 `@props` 태그를 임의로 지정해서 props 설명을 하는건가요??)

함수 컴포넌트의 경우 함수이니 `@param`으로 props를 설명하는 것을 본적은 있습니다!
만약 forwardRef로 감싸게 되면 또 어떻게 주석을 달 수 있는지 궁금하네요!

woo-jk replied:
props는 임의로 지정해줬어요~ 저렇게 써도 vscode에서는 잘 뜨긴 하더라구요 ㅋㅋ 근데 이렇게 임의로 써도 되는건진 확실하지 않네요.. 저도 좀 찾아봐야겠습니다! 일단은 param으로 쓰는게 좋아보여요

jariita replied:
컴포넌트가 ref를 지원하게 되면서 forwardRef로 감싸게 되었는데,
그렇게 되면 실제 param이 아니게 되는데 이럴때는 어떻게 해야될지 궁금하네요 🥲
지금은 문서를 주석으로 넣어놨습니다!

```tsx
/**
 * Docs:
 * - [Flex Docs](https://6507fe70aa120a5453701630-uaqgayiynj.chromatic.com/?path=/docs/example-flex--docs)
 */
```

woo-jk commented on src/components/Container.tsx:
```diff
@@ -0,0 +1,26 @@
+import { ComponentPropsWithoutRef, ElementType } from 'react';
+import { centerStyle } from '../styles/center';
+
+interface ContainerProps extends ComponentPropsWithoutRef<ElementType> {
```
ElementType을 상속해서 기존 요소들을 사용할 수 있도록 해준 점이 좋네요!

woo-jk commented on src/components/Container.tsx:
```diff
@@ -0,0 +1,26 @@
+  minWidth?: number;
+}
+
+function Container({ as = 'div', maxWidth, minWidth, children, ...attributes }: ContainerProps) {
```
optional props들에 기본 값을 지정해주지 않으면 undefined 값이 들어갈 것 같아요~!

jariita replied:
✍️ **[옵셔널 + 기본값]** - [ca5282f](https://github.com/woowacourse/layout-component/pull/43/commits/ca5282f89478b135eec1b2f40d423ce744939805)와 [0cad914](https://github.com/woowacourse/layout-component/pull/43/commits/0cad914be63a3cd2dcc7d2632eddd944be76bb04)에서 수정하였습니다!

woo-jk commented on src/components/Grid.tsx:
```diff
@@ -0,0 +1,59 @@
+   * `gap` 스타일 프로퍼티를 정의합니다.
+   * @default 0
+   */
+  gap: number;
```
이 부분도 optional이 README랑 다르네요!?

jariita replied:
✍️ **[옵셔널 + 기본값]** - [ca5282f](https://github.com/woowacourse/layout-component/pull/43/commits/ca5282f89478b135eec1b2f40d423ce744939805)와 [0cad914](https://github.com/woowacourse/layout-component/pull/43/commits/0cad914be63a3cd2dcc7d2632eddd944be76bb04)에서 수정하였습니다!

woo-jk commented on src/components/Grid.tsx:
```diff
@@ -0,0 +1,59 @@
+  const Component = as;
+
+  const gridTemplateRows = String(rows)?.match(/^\d+$/) ? `repeat(${rows}, minmax(0, 1fr))` : rows;
+  const gridTemplateColumns = String(columns)?.match(/^\d+$/) ? `repeat(${columns}, minmax(0, 1fr))` : columns;
```
이 로직들은 무엇을 의미하나요? 복잡한 로직에 참고할 주석이 있다면 좋을 것 같아요!

jariita replied:
✍️ 함수명으로 의도를 표현하도록 함수로 분리하였습니다! - [78dd65d](https://github.com/woowacourse/layout-component/pull/43/commits/78dd65d0701de1d3f4c2544ddea20681ebb71035)

woo-jk commented on src/stories/Container.stories.tsx:
```diff
@@ -0,0 +1,26 @@
+  component: Container,
+  parameters: {
+    docs: {
+      description: {
+        component: '화면의 폭에 따라 내용의 최대 폭을 제한하고 중앙에 배치하는 레이아웃을 만들기 위한 컴포넌트입니다.',
+      },
```
description 나이스입니다👍

woo-jk commented on webpack.config.js:
```diff
@@ -0,0 +1,30 @@
+  const isEnvDevelopment = webpackEnv === 'development';
+
+  return {
+    mode: isEnvDevelopment ? 'development' : 'production',
```
```
mode: webpackEnv
```

이렇게 되면 안되나요?!

jariita replied:
헛.. 그렇네요!
현재는 tsc로 빌드하고 있어서 webpack과 babel 설정을 제거하게 되었습니다!
(package.json에서는 아직 삭제가 안되었는데 step2 때는 패키지도 삭제해야되겠네요!)
tsc기 때문에 지금은 각 파일을 컴파일하고 있는데, rollup을 이용하면 cjs, esm 모두 대응하여 번들링 할 수 있더라구요!

woo-jk commented on src/styles/center.ts:
```diff
@@ -0,0 +1,5 @@
+import { css } from '@emotion/react';
+
+export const centerStyle = css({
+  margin: '0 auto',
+});
```
center 스타일을 추상화 시켜줬군요👍

woo-jk commented on package.json:
```diff
@@ -0,0 +1,61 @@
+{
+  "name": "conan-layout-component",
+  "version": "0.1.0",
+  "description": "Layout components for React",
+  "keywords": [
+    "react",
+    "layout"
+  ],
+  "homepage": "https://github.com/cruelladevil/layout-component",
+  "bugs": {
+    "url": "https://github.com/cruelladevil/layout-component/issues",
+    "email": "dev.timetravel@gmail.com"
+  },
+  "repository": {
+    "type": "git",
+    "url": "git+https://github.com/cruelladevil/layout-component.git"
+  },
+  "author": {
+    "name": "cruelladevil",
+    "email": "dev.timetravel@gmail.com",
+    "url": "https://github.com/cruelladevil"
+  },
```
package.json이 상세해서 좋아요👍

woo-jk reviewed (changes requested):
안녕하세요 코난👋
지난 미션에서 리뷰어로 만났는데 이번엔 리뷰이로 만나네요 ㄷㄷ

전반적으로 잘해주셔서 크게 바뀔 점은 없었고, 살짝 아쉬운 부분이나 궁금한 점에 대해 코멘트 남겼으니 답변 부탁드릴게요~! 고생하셨습니다~

### 레이아웃 컴포넌트가 여러 환경에서도 잘 보이는지 확인해 주세요
- [x] 반응형: 레이아웃이 다양한 화면 크기에 적응하는지 확인해 주세요.
- [x] 크로스 브라우징: 여러 브라우저(크롬, 사파리, 엣지 등)에서 동작하는지 확인해 주세요.
반응형에서도 잘 작동하고 여러 브라우저에서도 잘 작동했어요.

### 레이아웃 컴포넌트를 사용하는 사용자로서, 잘 이해하고 활용할 수 있었는지 피드백해 주세요
- [x] 가독성: README나 사용법을 보고, 저자의 의도가 잘 이해되는지 독자의 입장에서 확인해 주세요.
- [ ] 사용성: 레이아웃 컴포넌트를 사용하는 입장에서 쉽게 사용할 수 있었는지 확인해 주세요.
전반적으로 코드가 이해하기 쉬웠어요! 그러나 아쉬운 점이라면 README에서 명시된 props와 실제 props 타입이 맞지 않는 경우가 조금 있었어요. 코멘트도 남겼으니 확인 부탁드릴게요~!
또한 프롭스 타입별로 주석 달아주신 점도 좋았습니다. 이 부분에서도 역시 코멘트 드렸습니다~!

### 스타일 코드가 네이밍, 재사용성 등을 고려해 잘 작성되어 있는지 확인해 주세요
- [x] 의미론적 네이밍: 해당 CSS 클래스나 ID의 이름이 의미론적으로 적절한지 확인해 주세요.
- [x] 재사용성: 중복되거나 불필요한 스타일 규칙이 있는지 살펴보세요.
전반적으로 네이밍이나 재사용적인 측면에서 문제 되는 점은 없었어요~

jariita committed:
docs: optional props들 수정

jariita committed:
feat: optional 처리 및 default 값 추가

jariita committed:
feat: 다형성 컴포넌트를 위한 타입 정의

jariita committed:
feat: Container 컴포넌트 수정

- 다형성 컴포넌트 타입 적용
- style과 attribute 모두 수용하도록 수정
- CSSProperties를 기반으로 interface 수정

jariita committed:
feat: Grid 컴포넌트 수정

- 다형성 컴포넌트 타입 적용
- style과 attribute 모두 수용하도록 수정
- CSSProperties를 기반으로 interface 수정

jariita committed:
feat: Flex 컴포넌트 수정

- 다형성 컴포넌트 타입 적용
- style과 attribute 모두 수용하도록 수정
- CSSProperties를 기반으로 interface 수정

jariita committed:
refactor: grid 행과 열을 계산하는 함수 분리

jariita committed:
docs: README.md 업데이트

jariita committed:
remove: 사용되지 않는 webpack과 babel 설정 파일 삭제

jariita committed:
chore: esm과 cjs 모두 지원하도록 tsc로 컴파일

woo-jk reviewed (approved):
고생하셨어요~ 이만 머지하겠습니다~ 답변 안된 코멘트들은 나중에 답변 부탁드리겠습니다~!
